File: services/crypto-core/session.go
package cryptocore

import (
	"crypto/ed25519"
	"crypto/sha256"
	"errors"
	"io"

	"golang.org/x/crypto/curve25519"
	"golang.org/x/crypto/hkdf"
)

const hkdfInfoX3DH = "SecuMSG-X3DH"

// InitSession performs the X3DH handshake as the initiator using the remote
// prekey bundle and prepares the initial Double Ratchet state.
func (d *Device) InitSession(bundle *PrekeyBundle) (*SessionState, *HandshakeMessage, error) {
	if d == nil {
		return nil, nil, errors.New("cryptocore: nil device")
	}
	if bundle == nil {
		return nil, nil, errors.New("cryptocore: nil bundle")
	}
	if err := verifyPrekeyBundle(bundle); err != nil {
		return nil, nil, err
	}

	ephemeral, err := generateX25519KeyPair()
	if err != nil {
		return nil, nil, err
	}

	var otk *OneTimePrekey
	if len(bundle.OneTimePrekeys) > 0 {
		otk = &bundle.OneTimePrekeys[0]
	}

	secret, err := deriveSharedSecretInitiator(d, bundle, ephemeral, otk)
	if err != nil {
		return nil, nil, err
	}
	root, chain := deriveInitialKeys(secret)

	var pending *uint32
	if otk != nil {
		pending = new(uint32)
		*pending = otk.ID
	}

	sess := &SessionState{
		RootKey:         root,
		SendChain:       chainState{Key: chain},
		RecvChain:       chainState{},
		RatchetPrivate:  ephemeral.Private,
		RatchetPublic:   ephemeral.Public,
		RemoteRatchet:   bundle.SignedPrekey,
		RemoteIdentity:  bundle.IdentityKey,
		RemoteSignature: append([]byte(nil), bundle.IdentitySignatureKey...),
		Role:            RoleInitiator,
		PendingPrekey:   pending,
		skipped:         make(map[string][32]byte),
	}

	msg := &HandshakeMessage{
		IdentityKey:          d.identity.dhPublic,
		IdentitySignatureKey: append([]byte(nil), d.identity.signingPublic...),
		EphemeralKey:         ephemeral.Public,
		OneTimePrekeyID:      pending,
	}
	return sess, msg, nil
}

// AcceptSession finalizes the X3DH handshake on the responder side using the
// initiator's handshake message and prepares the receiving Double Ratchet state.
func (d *Device) AcceptSession(msg *HandshakeMessage) (*SessionState, error) {
	if d == nil {
		return nil, errors.New("cryptocore: nil device")
	}
	if msg == nil {
		return nil, errors.New("cryptocore: nil handshake message")
	}
	var otk *keyPair
	if msg.OneTimePrekeyID != nil {
		entry, ok := d.oneTime[*msg.OneTimePrekeyID]
		if !ok {
			return nil, ErrMissingOneTimeKey
		}
		k := entry.key
		otk = &k
		delete(d.oneTime, *msg.OneTimePrekeyID)
	}
	secret, err := deriveSharedSecretResponder(d, msg, otk)
	if err != nil {
		return nil, err
	}
	root, chain := deriveInitialKeys(secret)

	sess := &SessionState{
		RootKey:         root,
		SendChain:       chainState{},
		RecvChain:       chainState{Key: chain},
		RatchetPrivate:  d.signedPrekey.Private,
		RatchetPublic:   d.signedPrekey.Public,
		RemoteRatchet:   msg.EphemeralKey,
		RemoteIdentity:  msg.IdentityKey,
		RemoteSignature: append([]byte(nil), msg.IdentitySignatureKey...),
		Role:            RoleResponder,
		PendingPrekey:   msg.OneTimePrekeyID,
		skipped:         make(map[string][32]byte),
	}
	return sess, nil
}

func verifyPrekeyBundle(bundle *PrekeyBundle) error {
	if len(bundle.IdentitySignatureKey) != ed25519.PublicKeySize {
		return ErrInvalidPrekeySignature
	}
	if !ed25519.Verify(ed25519.PublicKey(bundle.IdentitySignatureKey), bundle.SignedPrekey[:], bundle.SignedPrekeySig) {
		return ErrInvalidPrekeySignature
	}
	return nil
}

func deriveSharedSecretInitiator(d *Device, bundle *PrekeyBundle, eph keyPair, otk *OneTimePrekey) ([]byte, error) {
	dh1, err := curve25519.X25519(d.identity.dhPrivate[:], bundle.SignedPrekey[:])
	if err != nil {
		return nil, err
	}
	dh2, err := curve25519.X25519(eph.Private[:], bundle.IdentityKey[:])
	if err != nil {
		return nil, err
	}
	dh3, err := curve25519.X25519(eph.Private[:], bundle.SignedPrekey[:])
	if err != nil {
		return nil, err
	}
	secret := append(append(append([]byte{}, dh1...), dh2...), dh3...)
	if otk != nil {
		dh4, err := curve25519.X25519(eph.Private[:], otk.Public[:])
		if err != nil {
			return nil, err
		}
		secret = append(secret, dh4...)
	}
	return secret, nil
}

func deriveSharedSecretResponder(d *Device, msg *HandshakeMessage, otk *keyPair) ([]byte, error) {
	dh1, err := curve25519.X25519(d.signedPrekey.Private[:], msg.IdentityKey[:])
	if err != nil {
		return nil, err
	}
	dh2, err := curve25519.X25519(d.identity.dhPrivate[:], msg.EphemeralKey[:])
	if err != nil {
		return nil, err
	}
	dh3, err := curve25519.X25519(d.signedPrekey.Private[:], msg.EphemeralKey[:])
	if err != nil {
		return nil, err
	}
	secret := append(append(append([]byte{}, dh1...), dh2...), dh3...)
	if otk != nil {
		dh4, err := curve25519.X25519(otk.Private[:], msg.EphemeralKey[:])
		if err != nil {
			return nil, err
		}
		secret = append(secret, dh4...)
	}
	return secret, nil
}

func deriveInitialKeys(secret []byte) ([32]byte, [32]byte) {
	kdf := hkdf.New(sha256.New, secret, nil, []byte(hkdfInfoX3DH))
	var root, chain [32]byte
	if _, err := io.ReadFull(kdf, root[:]); err != nil {
		// return zero-value keys on failure
		return [32]byte{}, [32]byte{}
	}
	if _, err := io.ReadFull(kdf, chain[:]); err != nil {
		// return zero-value keys on failure
		return [32]byte{}, [32]byte{}
	}
	return root, chain
}

File: frontend/src/crypto-core/session.ts
import { ed25519 } from "@noble/curves/ed25519";
import { curve25519 } from "@noble/curves/curve25519";
import { hkdf } from "@noble/hashes/hkdf";
import { sha256 } from "@noble/hashes/sha256";
import { Device, generateX25519KeyPair } from "./core";
import { ErrInvalidPrekeySignature, ErrMissingOneTimeKey } from "./errors";
import type {
  ChainState,
  HandshakeMessage,
  OneTimePrekey,
  PrekeyBundle,
  SessionRole,
  SessionState,
} from "./types";
import { copyBytes, concatBytes, utf8, zeroBytes } from "./utils";

const hkdfInfoX3DH = "SecuMSG-X3DH";

export function InitSession(
  d: Device,
  bundle: PrekeyBundle,
): { session: SessionState; message: HandshakeMessage } {
  if (!d) {
    throw new Error("cryptocore: nil device");
  }
  if (!bundle) {
    throw new Error("cryptocore: nil bundle");
  }
  verifyPrekeyBundle(bundle);

  const ephemeral = generateX25519KeyPair();
  const otk = bundle.OneTimePrekeys.length > 0 ? bundle.OneTimePrekeys[0] : undefined;
  const secret = deriveSharedSecretInitiator(d, bundle, ephemeral, otk);
  const { root, chain } = deriveInitialKeys(secret);

  const pending = otk ? otk.ID : undefined;

  const session: SessionState = {
    RootKey: root,
    SendChain: { Key: chain, Index: 0 },
    RecvChain: { Key: zeroBytes(32), Index: 0 },
    RatchetPrivate: copyBytes(ephemeral.Private),
    RatchetPublic: copyBytes(ephemeral.Public),
    RemoteRatchet: copyBytes(bundle.SignedPrekey),
    RemoteIdentity: copyBytes(bundle.IdentityKey),
    RemoteSignature: copyBytes(bundle.IdentitySignatureKey),
    PN: 0,
    Role: SessionRole.RoleInitiator,
    PendingPrekey: pending,
    skipped: new Map(),
  };

  const msg: HandshakeMessage = {
    IdentityKey: copyBytes(d.identity.dhPublic),
    IdentitySignatureKey: copyBytes(d.identity.signingPublic),
    EphemeralKey: copyBytes(ephemeral.Public),
    OneTimePrekeyID: pending,
  };

  return { session, message: msg };
}

export function AcceptSession(d: Device, msg: HandshakeMessage): SessionState {
  if (!d) {
    throw new Error("cryptocore: nil device");
  }
  if (!msg) {
    throw new Error("cryptocore: nil handshake message");
  }
  let otk: { Private: Uint8Array; Public: Uint8Array } | undefined;
  if (typeof msg.OneTimePrekeyID === "number") {
    const entry = d.oneTime.get(msg.OneTimePrekeyID);
    if (!entry) {
      throw ErrMissingOneTimeKey;
    }
    otk = entry.key;
    d.oneTime.delete(msg.OneTimePrekeyID);
  }
  const secret = deriveSharedSecretResponder(d, msg, otk);
  const { root, chain } = deriveInitialKeys(secret);

  const session: SessionState = {
    RootKey: root,
    SendChain: { Key: zeroBytes(32), Index: 0 },
    RecvChain: { Key: chain, Index: 0 },
    RatchetPrivate: copyBytes(d.signedPrekey.Private),
    RatchetPublic: copyBytes(d.signedPrekey.Public),
    RemoteRatchet: copyBytes(msg.EphemeralKey),
    RemoteIdentity: copyBytes(msg.IdentityKey),
    RemoteSignature: copyBytes(msg.IdentitySignatureKey),
    PN: 0,
    Role: SessionRole.RoleResponder,
    PendingPrekey: msg.OneTimePrekeyID,
    skipped: new Map(),
  };
  return session;
}

function verifyPrekeyBundle(bundle: PrekeyBundle): void {
  if (bundle.IdentitySignatureKey.length !== 32) {
    throw ErrInvalidPrekeySignature;
  }
  const valid = ed25519.verify(
    bundle.SignedPrekeySig,
    bundle.SignedPrekey,
    bundle.IdentitySignatureKey,
  );
  if (!valid) {
    throw ErrInvalidPrekeySignature;
  }
}

function deriveSharedSecretInitiator(
  d: Device,
  bundle: PrekeyBundle,
  eph: ReturnType<typeof generateX25519KeyPair>,
  otk?: OneTimePrekey,
): Uint8Array {
  const dh1 = curve25519.scalarMult(d.identity.dhPrivate, bundle.SignedPrekey);
  const dh2 = curve25519.scalarMult(eph.Private, bundle.IdentityKey);
  const dh3 = curve25519.scalarMult(eph.Private, bundle.SignedPrekey);
  let secret = concatBytes(dh1, dh2, dh3);
  if (otk) {
    const dh4 = curve25519.scalarMult(eph.Private, otk.Public);
    secret = concatBytes(secret, dh4);
  }
  return secret;
}

function deriveSharedSecretResponder(
  d: Device,
  msg: HandshakeMessage,
  otk?: { Private: Uint8Array; Public: Uint8Array },
): Uint8Array {
  const dh1 = curve25519.scalarMult(d.signedPrekey.Private, msg.IdentityKey);
  const dh2 = curve25519.scalarMult(d.identity.dhPrivate, msg.EphemeralKey);
  const dh3 = curve25519.scalarMult(d.signedPrekey.Private, msg.EphemeralKey);
  let secret = concatBytes(dh1, dh2, dh3);
  if (otk) {
    const dh4 = curve25519.scalarMult(otk.Private, msg.EphemeralKey);
    secret = concatBytes(secret, dh4);
  }
  return secret;
}

function deriveInitialKeys(secret: Uint8Array): { root: ChainState["Key"]; chain: ChainState["Key"]; } {
  const okm = hkdf(sha256, secret, new Uint8Array(), utf8(hkdfInfoX3DH), 64);
  return {
    root: okm.slice(0, 32),
    chain: okm.slice(32, 64),
  };
}
